### ISSUE TITLE
Encrypting large files

### ISSUE THREAD

**Issue by ckcr4lyf** [sh-dv/hat.sh#11]
I came across this projecting when looking at how to implement very large file encryption (2GB+) client side, in the browser.

When I loaded a 3.5GB mkv and clicked "encrypt", nothing happened.  (Tested on latest Google Chrome, Win 10, 8GB RAM). From the code, it seems the file is loaded in memory and encrypted.

Do you have an idea for a solution to encrypt files of arbitrary size?

**Comment by httpjamesm** [comment]
I also have this issue.

**Comment by lorenzo95** [comment]
same here. Nothing happens.

**Comment by ckcr4lyf** [comment]
Well, it seems this is indeed a browser limitation. What I can suggest you guys @httpjamesm @lorenzo95 do is use a local program, such a VeraCrypt, for encryption instead.

**Comment by th3wach3r** [comment]
actually guys, sh-dv is already working on v2.0, which will support unlimited file size encryption.
the current code works, but the thing is it loads the whole file into memory and encrypts it. which results in the browser crashing if you use a file larger than 2.5gb (talking 8gb ram). be patient.
he clearly needs support to be able to continue working on this, and he got nothing. 
v2 is coming, me and him will publish another project that uses hat.sh also.
“Love and hate are two sides of the same coin.”.
peace.

**Comment by ckcr4lyf** [comment]
Don't worry, I dont think any of us intended to blame anyone, rather I was actually researching into the problem when I came across this service. 
I'd be happy to help, but from my current understanding, its a browser based limitation. I don't think currently the browsers allow buffering a file partially by specify the bytes to read specifically?

**Comment by th3wach3r** [comment]
exactly, the browsers have limitations, but the way to bypass those limitation is by chunking the file to small size chunks before the file is read by the server. for example taking a 1 GB file and slice it to 100 bytes chunks. and encrypt each chunk and then merge them as one file, and the same thing goes for decryption. AES-GCM supports this method. but there is a downside to that, in order to determine the sequence number of each chunk, you have to store information(nonce+sequnce nr) in each chunk, this will take also some bytes size . then, if you are going to encrypt a 1GB file , you are going to end up with 1.3GB encrypted file. because it contains those infos.

**Comment by th3wach3r** [comment]
exactly, the browsers have limitations, but the way to bypass those limitation is by chunking the file to small size chunks before the file is read by the browser. for example taking a 1 GB file and slice it to 100 bytes chunks. and encrypt each chunk and then merge them as one file, and the same thing goes for decryption. AES-GCM supports this method. but there is a downside to that, in order to determine the sequence number of each chunk, you have to store information(nonce+sequnce nr) in each chunk, this will take also some bytes size . then, if you are going to encrypt a 1GB file , you are going to end up with 1.3GB encrypted file. because it contains those infos.

**Comment by ckcr4lyf** [comment]
For encryption I guess this would work, as while reading it is possible to seek to certain offsets and read a particular sequence of bytes. But when decrypting (writing), I dont think browsers give enough access to the filesystem to allow seeking to a particular point and writing data in the filesystem, and the blob size would be limited by memory?
